src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,315 @@
+    const eventCreateInputs = events.map(this.stringifyEvent);
+
+    let stream = await this.collection.findOne({
+      streamName: { $eq: streamName },
+    });
+    let createdNewStream = false;
+
+    if (!stream) {
+      const result = await this.collection.insertOne({
+        streamName,
+        events: [],
+        createdAt: new Date(),
+        updatedAt: new Date(),
+      });
+      stream = await this.collection.findOne({
+        _id: result.insertedId,
+      });
+      createdNewStream = true;
+    }
+
+    // TODO: better error here, should rarely happen if ever
+    // if another error was not thrown before this
+    if (!stream) throw new Error('Failed to create stream');
+
+    assertExpectedVersionMatchesCurrent(
+      stream.events.length,
+      options?.expectedStreamVersion,
+      MongoDBEventStoreDefaultStreamVersion,
+    );
+
+    const updatedStream = await this.collection.findOneAndUpdate(
+      {
+        streamName: { $eq: streamName },
+        events: { $size: stream.events.length },
+        updatedAt: new Date(),
+      },
+      { $push: { events: { $each: eventCreateInputs } } },
+      { returnDocument: 'after' },
+    );
+
+    if (!updatedStream) {
+      const currentStream = await this.collection.findOne({
+        streamName: { $eq: streamName },
+      });
+      throw new ExpectedVersionConflictError(
+        currentStream?.events.length ?? -1,
+        stream.events.length,
+      );
+    }
+
+    const { streamType, entityId } = fromStreamName(streamName);
+
+    if (options?.projections) {
+      await Promise.all(
+        options.projections.map((project) =>
+          project({
+            streamName,
+            streamType,
+            entityId,
+            streamVersion: updatedStream.events.length,
+            events: updatedStream.events.map(this.parseEvent(streamName)),
+          }),
+        ),
+      );
+    }
+
+    if (options?.asyncProjections) {
+      for (const project of options.asyncProjections) {
+        project({
+          streamName,
+          streamType,
+          entityId,
+          streamVersion: updatedStream.events.length,
+          events: updatedStream.events.map(this.parseEvent(streamName)),
+        });
+      }
+    }
+
+    return {
+      nextExpectedStreamVersion: updatedStream.events.length,
+      createdNewStream,
+    };
+  }
+
+  /**
+   * Transforms the `event` from the saved format into the usable object
+   * at runtime. This function may be altered later to match `stringifyEvent`.
+   */
+  private parseEvent<EventType extends Event>(streamName: StreamName) {
+    return (
+      event: EventStreamEvent,
+      index?: number,
+    ): ReadEvent<EventType, ReadEventMetadata> => {
+      const metadata = {
+        ...JSON.parse(event.metadata),
+        eventId: event._id,
+        streamName,
+        streamPosition: BigInt(index ?? 0),
+      } satisfies ReadEventMetadata;
+
+      // TODO: is this correct?
```

Comment:
I _think_ this is ok, TS just doesn't like it.

Reply:
Is it outdated?

Reply:
Yes